identity-iota/src/credential/credential_validator.rs

```diff
@@ -0,0 +1,645 @@
+    Self {}
+  }
+
+  /// Validates the semantic structure of the [Credential].
+  ///
+  /// # Terminology
+  /// This is a *validation unit*
+  pub fn check_structure<T>(credential: &Credential<T>) -> Result<()> {
+    Self::check_structure_local_error(credential).map_err(Error::UnsuccessfulValidationUnit)
+  }
+
+  fn check_structure_local_error<T>(credential: &Credential<T>) -> ValidationUnitResult {
+    credential
+      .check_structure()
+      .map_err(ValidationError::CredentialStructure)
+  }
+
+  /// Validate that the [Credential] does not expire before the specified [Timestamp].
+  ///
+  /// # Terminology
+  /// This is a *validation unit*
+  pub fn does_not_expire_before<T>(credential: &Credential<T>, timestamp: Timestamp) -> Result<()> {
+    Self::does_not_expire_before_local_error(credential, timestamp).map_err(Error::UnsuccessfulValidationUnit)
+  }
+
+  fn does_not_expire_before_local_error<T>(credential: &Credential<T>, timestamp: Timestamp) -> ValidationUnitResult {
+    let is_ok = if let Some(expiration_date) = credential.expiration_date {
+      expiration_date >= timestamp
+    } else {
+      true
+    };
+    is_ok.then(|| ()).ok_or(ValidationError::ExpirationDate)
+  }
+
+  /// Validate that the [Credential] is issued no later than the specified [Timestamp].
+  ///
+  /// # Terminology
+  /// This is a *validation unit*
+  pub fn was_not_issued_after<T>(credential: &Credential<T>, timestamp: Timestamp) -> Result<()> {
+    Self::was_not_issued_after_local_error(credential, timestamp).map_err(Error::UnsuccessfulValidationUnit)
+  }
+
+  fn was_not_issued_after_local_error<T>(credential: &Credential<T>, timestamp: Timestamp) -> ValidationUnitResult {
+    (credential.issuance_date <= timestamp)
+      .then(|| ())
+      .ok_or(ValidationError::IssuanceDate)
+  }
+
+  // helper method for verifying the signature of a trusted issuer
+  fn extract_corresponding_issuer_helper<'a, T, I: Iterator<Item = &'a ResolvedIotaDocument>>(
+    credential: &Credential<T>,
+    mut issuer_docs: I,
+  ) -> std::result::Result<&'a ResolvedIotaDocument, ValidationError> {
+    let issuer_did: Result<IotaDID> = credential.issuer.url().as_str().parse();
+    match issuer_did {
+      Ok(did) => {
+        // if the issuer_did corresponds to one of the trusted issuers we use the corresponding DID Document to verify
+        // the signature
+        issuer_docs
+          .find(|issuer_doc| issuer_doc.document.id() == &did)
+          .ok_or(ValidationError::IncompatibleIssuerDocuments)
+      }
+      Err(error) => {
+        // the issuer's url could not be parsed to a valid IotaDID
+        Err(ValidationError::IssuerUrl { source: error.into() })
+      }
+    }
+  }
+  /// Verify the signature using the DID Document of a trusted issuer.
+  ///
+  /// This method will only attempt to verify the credential's signature if
+  /// the credential issuer url can be parsed to a DID belonging to one of the trusted issuers.
+  ///
```

Comment:
Nit: maybe clarify that it returns an error, the wording here might imply that verification is skipped in such a case.

Aside: in general, do we want to enumerate error cases in the doc-comments like in other places?

Reply:
>Nit: maybe clarify that it returns an error, the wording here might imply that verification is skipped in such a case.

I introduced an # Errors section where this is explained. 

> Aside: in general, do we want to enumerate error cases in the doc-comments like in other places? 

I think we should do it when we expect  the caller to wonder about what the error conditions may be, but it does not have to be applied universally.